04.md

```diff
@@ -40,13 +44,50 @@ let event = {
   created_at: Math.floor(Date.now() / 1000),
   kind: 4,
   tags: [['p', theirPublicKey]],
-  content: encryptedMessage + '?iv=' + ivBase64
+  content: encryptedMessage + '??' + ivBase64 + "??1"
+}
+```
+
+
+Code sample for decrypting both version 0 and version 1 events in JavaScript:
+
+```js
+import crypto from 'crypto'
+import * as secp from '@noble/secp256k1'
+import { hkdf } from '@noble/hashes/hkdf';
+
+function trimEnd(str, ch) {
+    var start = 0, 
+        end = str.length;
+    while(end > 0 && str[end - 1] === ch)
+        --end;
+    return (end < str.length) ? str.substring(0, end) : str;
 }
+
+let [ctb64, ivb64, version] = data.split('??')
+if (version != "1")
+    [ctb64, ivb64] = data.split('?iv=')
+let iv = base64.decode(ivb64)
+let sharedPoint = secp.getSharedSecret(ourPrivateKey, '02' + theirPublicKey)
+let sharedX = sharedPoint.slice(1, 33)
+let key
+if (version == "1")
+    key = hkdf(sha256, normalizedKey, iv, undefined, 32)
```

Comment:
That's false. It's 10 times slower than sha256.

Reply:
indeed, it's slower unless you're careful to concatenate the buffers correctly.

Reply:
test: 2070 ms version 0
      test: 2969 ms version 1 sha256
      test: 3043 ms version 1 hkdf
      
      hkdf is negligible and it's best practice.   lets stick with it!

Reply:
honestly, if speed is a concern, i'd rather have key-rotation and cbc, than no key rotation and gcm

Reply:
>no, it's not safe to reuse the key if the iv is short. reusing the same iv allows the attacker to decrypt messages, and there is a birthday attack here.

So just use xchacha instead of these "constructions".